File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
while True:
    li = []
    state = True
    S = input()
    if S == '.':
        break
    for s in S:
        if s == '(' or s == '[':
            li.append(s)
        elif s == ')':
            if li and li[-1] == '(':
                li.pop()
            else:
                state = False
                break
        elif s == ']':
            if li and li[-1] == '[':
                li.pop()
            else:
                state = False
                break

    if state and not li:
        print('yes')
    else:
        print('no')

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <stack>


using namespace std;


int main() {
    ios::sync_with_stdio(false);
    ios_base::sync_with_stdio(0);
    cin.tie(0);


    while(true){
        string input;
        getline(cin, input);
        if(input==".")
            break;

        stack<char> st;
        string val="yes";

        for(int i=0;i<input.length();i++){
            if((input[i]=='(')||(input[i]=='[')){
                st.push(input[i]);
            }
            else if(input[i]==')'){
                if(!st.empty()&&st.top()=='('){
                    st.pop();
                }
                else{
                    val="no";
                    break;
                }
            }
            else if(input[i]==']'){
                if(!st.empty()&&st.top()=='['){
                    st.pop();
                }
                else{
                    val="no";
                    break;
                }
            }
        }
        if(st.empty()&&val=="yes"){
            cout<<"yes"<<'\n';
        }
        else{
            cout<<"no"<<'\n';
        }
    }
}




